### PR TITLE
Allow control of --check-bounds command option

### PR DESCRIPTION
**Pull request by IanButterworth**
I believe this command option is a bit more complicated than setting `--check-bounds` to yes (force always) or no (force never), because the default state is a 3rd state where `--check-bounds` is omitted entirely, and bounds checking is done per code decisions.. It'd be good if someone could confirm my understanding on that though.. the code is a bit hard to follow.

This is untested. I've not figured out how to test this in an action yet

### CODE REVIEW

**Comment by IanButterworth** [comment]
I'm not sure how to fix the bash here. Hopefully it's clear what i'm trying to do

**Comment by IanButterworth** [comment]
Ok. Should be fixed now

**Comment by DilumAluthge** [comment]
Can you squash the commits?

At a glance, it looks right to me, but I'm not the most familiar with this repo, so it would be good to get an approval from someone that knows the code better than I do.

**Comment by SaschaMann** [comment]
I'll take a look later. I can squash it on merge, no need for it now.

File: action.yml [hunk 1]
```diff
@@ -48,7 +51,12 @@ runs:
         JULIA_PKG_SERVER: ""
     - run: |
         # The Julia command that will be executed
-        julia_cmd=( julia --check-bounds=yes --color=yes --depwarn=${{ inputs.depwarn }} --inline=${{ inputs.inline }} --project=${{ inputs.project }} -e 'import Pkg;include(joinpath(ENV["GITHUB_ACTION_PATH"], "kwargs.jl"));kwargs = Kwargs.kwargs(;coverage = :(${{ inputs.coverage }}),force_latest_compatible_version = :(${{ inputs.force_latest_compatible_version }}),);Pkg.test(; kwargs...)' )
+        check_bounds_opt=' '
+        if [[ ${{ inputs.check_bounds }} != 'auto' ]]
+        then
+          check_bounds_opt='--check-bounds=${{ inputs.check_bounds }} '
+        fi
+        julia_cmd=( julia ${check_bounds_opt}--color=yes --depwarn=${{ inputs.depwarn }} --inline=${{ inputs.inline }} --project=${{ inputs.project }} -e 'import Pkg;include(joinpath(ENV["GITHUB_ACTION_PATH"], "kwargs.jl"));kwargs = Kwargs.kwargs(;coverage = :(${{ inputs.coverage }}),force_latest_compatible_version = :(${{ inputs.force_latest_compatible_version }}),);Pkg.test(; kwargs...)' )
```
Comment (SaschaMann) [inline action.yml:59]:
```suggestion
        julia_cmd=( julia "$check_bounds_opt" --color=yes --depwarn=${{ inputs.depwarn }} --inline=${{ inputs.inline }} --project=${{ inputs.project }} -e 'import Pkg;include(joinpath(ENV["GITHUB_ACTION_PATH"], "kwargs.jl"));kwargs = Kwargs.kwargs(;coverage = :(${{ inputs.coverage }}),force_latest_compatible_version = :(${{ inputs.force_latest_compatible_version }}),);Pkg.test(; kwargs...)' )
```

Shellcheck complains about the lack of quotes here.

I feel like there's probably a better way to do add an argument like this, but I don't know bash well enough.

**Review by SaschaMann — COMMENTED** [review]
The script looks alright, but I'm not sure about the behaviour of bounds checks.

Could you perhaps run the action thrice in a row for a package/script with all three inputs to see if it fails/behaves as expected as a demo?

Perhaps in a fork of https://github.com/julia-actions/Example.jl/blob/action-tests/test/runtests.jl?

**Comment by IanButterworth** [comment]
Bounds check behavior confirmed on slack to be 

- `--check-bounds=yes`: force always bounds check
- `--check-bounds=no`: force never bounds check
- no command opt: automatic based on code controls


Happy to test run this with each config. How can I use this PR of the action though?

File: action.yml [hunk 2]
```diff
@@ -48,7 +51,12 @@ runs:
         JULIA_PKG_SERVER: ""
     - run: |
         # The Julia command that will be executed
-        julia_cmd=( julia --check-bounds=yes --color=yes --depwarn=${{ inputs.depwarn }} --inline=${{ inputs.inline }} --project=${{ inputs.project }} -e 'import Pkg;include(joinpath(ENV["GITHUB_ACTION_PATH"], "kwargs.jl"));kwargs = Kwargs.kwargs(;coverage = :(${{ inputs.coverage }}),force_latest_compatible_version = :(${{ inputs.force_latest_compatible_version }}),);Pkg.test(; kwargs...)' )
+        check_bounds_opt=' '
+        if [[ ${{ inputs.check_bounds }} != 'auto' ]]
+        then
+          check_bounds_opt='--check-bounds=${{ inputs.check_bounds }} '
+        fi
```
Comment (SaschaMann) [inline action.yml:58]: Sorry for causing that error, bash quotes are hard 🤦 

```suggestion
        check_bounds_opt=''
        if [[ ${{ inputs.check_bounds }} != 'auto' ]]
        then
          check_bounds_opt='--check-bounds=${{ inputs.check_bounds }}'
        fi
```

Maybe this? If not, I'm fine with reverting it to what it was before since that didn't fail the build.

**Review by SaschaMann — COMMENTED** [review]
> Happy to test run this with each config. How can I use this PR of the action though?

Replace `- uses: julia-actions/julia-runtest@v1` with `- uses: IanButterworth/julia-runtest@ib/bounds_check_control` in the workflow in the repo/package you want to run the tests in. Doesn't have to be merged, you can replace and run it on a branch.

**Comment by IanButterworth** [comment]
Needs a fix in Pkg, as it turns out `--check-bounds=yes` is hardcoded into the Pkg test sandbox https://github.com/JuliaLang/Pkg.jl/pull/2668 and there's no way to override it back to default currently.

Would be easier with https://github.com/JuliaLang/julia/pull/41551

**Comment by SaschaMann** [comment]
Let's wait for that to be resolved then

**Comment by IanButterworth** [comment]
This PR https://github.com/IanButterworth/SystemBenchmark.jl/pull/50 is running on [`1`, `nightly`] with each of the check bounds options.

On nightly it is respected, and can be seen in the command args shown at the start of the test suite.

However on `1` (1.6) given https://github.com/JuliaLang/Pkg.jl/pull/2439 wasn't backported to 1.6, `--check-bounds=yes` cannot be overridden, but the tooling here doesn't error
Backporting that PR for this alone is probably not the best course of action, so perhaps this is just for 1.7 and up

File: kwargs.jl [hunk 3]
```diff
@@ -5,14 +5,21 @@ import Pkg
...
     end
 
     kwargs_dict = Dict{Symbol, Any}()
     kwargs_dict[:coverage] = coverage
 
+    if VERSION >= v"1.6.0"
+        kwargs_dict[:julia_args] = julia_args
+    elseif julia_args != ``
+        @warn("The `julia_args` option requires at least Julia 1.6", VERSION, julia_args)
+    end
```
Comment (SaschaMann) [inline kwargs.jl:21]:
```suggestion
    elseif julia_args != ``
        println("::warning::The `julia_args` option requires at least Julia 1.6. VERSION=$VERSION, julia_args=$julia_args")
    end
```

I think this will more noticable if it's a proper GHA-style warning. Doesn't seem worth it to install the GHA.jl package solely for the logger, so I'm using the [warning syntax](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message) manually.

**Comment by IanButterworth** [comment]
@SaschaMann Should be ready to merge as this works on nightly, and the backports of JuliaLang/julia#41551 are labelled

File: kwargs.jl [hunk 4]
```diff
@@ -5,14 +5,21 @@ import Pkg
...
     end
 
     kwargs_dict = Dict{Symbol, Any}()
     kwargs_dict[:coverage] = coverage
 
+    if VERSION >= v"1.6.0"
+        kwargs_dict[:julia_args] = julia_args
+    elseif julia_args != ``
+        println("::warning::The `julia_args` option requires at least Julia 1.6. VERSION=$VERSION, julia_args=$julia_args")
+    end
```
Comment (SaschaMann) [inline kwargs.jl:21]: Sorry to ask again: Can `julia_args` ever be ` `` ` here? It's set to `julia_args = ["--check-bounds=${{ inputs.check_bounds }}"]` in `action.yml` and `${{ inputs.check_bounds }}` defaults to `yes`.

Comment (IanButterworth) [reply kwargs.jl]: Good catch. Should be fixed now. The change makes kwargs.jl more specific to how it's used in this action, which I assumed wasn't a problem

**Comment by SaschaMann** [comment]
Sorry, I haven't had time to look at this again yet, it's on my list for the weekend.

File: kwargs.jl [hunk 5]
```diff
@@ -5,14 +5,23 @@ import Pkg
 include(joinpath(@__DIR__, "autodetect-dependabot.jl"))
 
 function kwargs(; coverage::Bool,
-                  force_latest_compatible_version::Union{Bool, Symbol})
+                  force_latest_compatible_version::Union{Bool, Symbol},
+                  julia_args::Union{Cmd, AbstractVector{<:AbstractString}}=``)
```
Comment (c42f) [inline kwargs.jl:9]: I guess `Cmd` isn't used in practice as `kwargs` is only called from the driver script?

```suggestion
                  julia_args::AbstractVector{<:AbstractString}}=String[])
```

**Review by c42f — APPROVED** [review]
I needed this today, so I had a read through the code.

Looks good to me!

File: kwargs.jl (hunk 5 above)
Comment (IanButterworth) [reply kwargs.jl:9]: Seems so! Sorry for the delay

**Comment by IanButterworth** [comment]
It'd be nice to get this in

File: kwargs.jl [hunk 6]
```diff
@@ -5,14 +5,23 @@ import Pkg
 include(joinpath(@__DIR__, "autodetect-dependabot.jl"))
 
 function kwargs(; coverage::Bool,
-                  force_latest_compatible_version::Union{Bool, Symbol})
+                  force_latest_compatible_version::Union{Bool, Symbol},
+                  julia_args::AbstractVector{<:AbstractString}}=String[])
```
Comment (IanButterworth) [inline kwargs.jl:9]:
```suggestion
                  julia_args::AbstractVector{<:AbstractString}=String[])
```

File: kwargs.jl (hunk 5 above)
Comment (c42f) [reply kwargs.jl]: All good!  I must admit, I forgot about this completely and can't even remember which package I wanted it for :-)

**Review by SaschaMann — APPROVED** [review]

**Comment by SaschaMann** [comment]
Sorry for the delay
